Make torch.library.define consistent with the new APIs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111310
* #111380
* #111309
* #111308
* __->__ #111307

This PR introduces a new overload of torch.library.define. Like
impl_abstract, and our plans for the rest of the torch.library APIs, we
allow it to accept an optional library object to tie the lifetime of the
op definition to.

Test Plan:
- new tests

zou3519 committed:
Make torch.library.define consistent with the new APIs

This PR introduces a new overload of torch.library.define. Like
impl_abstract, and our plans for the rest of the torch.library APIs, we
allow it to accept an optional library object to tie the lifetime of the
op definition to.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Make torch.library.define consistent with the new APIs"


This PR introduces a new overload of torch.library.define. Like
impl_abstract, and our plans for the rest of the torch.library APIs, we
allow it to accept an optional library object to tie the lifetime of the
op definition to.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Make torch.library.define consistent with the new APIs"


This PR introduces a new overload of torch.library.define. Like
impl_abstract, and our plans for the rest of the torch.library APIs, we
allow it to accept an optional library object to tie the lifetime of the
op definition to.

Test Plan:
- new tests

[ghstack-poisoned]

soulitzer commented on torch/_library/utils.py:
```diff
@@ -37,3 +37,15 @@ def get_source(stacklevel: int) -> str:
+    splits = name.split("::")
+    if len(splits) != 2:
+        raise ValueError(
+            f"Expected `name` to be of the form "
```
would be nice to name this `qualified_name` just to avoid the slight name = namespace::name confusion, but thats probably too bc-breaking

zou3519 replied:
It's not BC-breaking. We haven't included impl_abstract in a release yet. I can update all of these to qualified_name for less confusion.

zou3519 replied:
Going to do `qualname` for short unless we want to be verbose

soulitzer commented on torch/_library/utils.py:
```diff
@@ -37,3 +37,15 @@ def get_source(stacklevel: int) -> str:
+        raise ValueError(
+            f"Expected `name` to be of the form "
+            f'"namespace::name", but got {name}. '
+            f"Operator names in PyTorch consist of a name "
```
it still feels like "Operator names in PyTorch" are still things like sin/cos?
when you actually mean "`name` passed to define/impl_* APIs consist of..."

zou3519 replied:
Yup

soulitzer commented on torch/library.py:
```diff
@@ -175,7 +176,68 @@ def wrap(f):
+    lib.define(name + schema, alias_analysis="")
+
+
+@define.register
```
woah, pretty cool. 
Does this use type annotations underneath?
And if someone passes in a torch.Library as the first argument, does it know to prioritize this old impl because it is annotated and the new API's first arg isn't annotated?

zou3519 replied:
functools.single_dispatch switches on the type of the first argument. If the type of the first argument is a Library, we got to the legacy overload. Otherwise, it hits the new function we've created here.

soulitzer reviewed (approved):
Sounds good, just some small nits

Is the idea behind the new APIs to abstract the library away from the user, and the user only deals with namespaces?

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -1675,6 +1675,33 @@ def foo_backward(ctx, saved, grad_out):
+        def f(x):
+            return torch.from_numpy(np.sin(x.numpy()))
+
+        lib.impl("foo", f, "CPU")
```
Decorator style is supported for impl. Decorator style is not supported for define. This is because there is no function to actually decorate for define.

zou3519 replied:
In the next PR up you will see decorator support for impl in the tests.

zou3519 committed:
Update on "Make torch.library.define consistent with the new APIs"


This PR introduces a new overload of torch.library.define. Like
impl_abstract, and our plans for the rest of the torch.library APIs, we
allow it to accept an optional library object to tie the lifetime of the
op definition to.

Test Plan:
- new tests

[ghstack-poisoned]
